contracts/voting-verifier/src/execute.rs

```diff
@@ -88,16 +94,66 @@ pub fn verify_messages(
 }
 
 pub fn vote(
-    _deps: DepsMut,
-    _info: MessageInfo,
-    _poll_id: String,
-    _votes: Vec<bool>,
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    poll_id: PollID,
+    votes: Vec<bool>,
 ) -> Result<Response, ContractError> {
-    todo!()
+    let mut poll = POLLS
+        .may_load(deps.storage, poll_id)?
+        .ok_or(ContractError::PollNotFound {})?;
+
+    poll.cast_vote(env.block.height, &info.sender, &votes)?;
+    POLLS.save(deps.storage, poll_id, &poll)?;
+
+    Ok(Response::new().add_event(
+        Voted {
+            poll_id,
+            voter: info.sender,
+        }
+        .into(),
+    ))
 }
 
-pub fn end_poll(_deps: DepsMut, _poll_id: String) -> Result<Response, ContractError> {
-    todo!()
+pub fn end_poll(deps: DepsMut, env: Env, poll_id: PollID) -> Result<Response, ContractError> {
+    let mut poll = POLLS
+        .may_load(deps.storage, poll_id)?
+        .ok_or(ContractError::PollNotFound {})?;
+
+    let poll_result = poll.tally(env.block.height)?;
+    POLLS.save(deps.storage, poll_id, &poll)?;
+
+    let messages = load_and_remove_pending_message(deps.storage, poll_id)?;
+    if messages.len() != poll_result.results.len() {
+        panic!(
+            "poll {} results and pending messages have different length",
+            poll_id
+        )
+    }
+
+    let messages = messages
+        .iter()
+        .zip(poll_result.results.iter())
+        .filter_map(|(message, verified)| match *verified {
+            true => Some(message),
+            false => None,
+        })
+        .collect::<Vec<&Message>>();
+
+    for message in messages {
+        VERIFIED_MESSAGES.save(deps.storage, &message.id, message)?;
+    }
```

Comment:
Do we need to check for message mismatch here? What if I create two polls at the same time, with same message ids but different messages?